ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_dev/ExplicitDisclosureIT.scala

```diff
@@ -0,0 +1,753 @@
+        streamType = "activeContracts",
+        expectedCount = 2,
+        streamResponses = acs,
+        toCreateEvents = Vector(_),
+        // ACS does not have effectiveAt, so use the one from the other streams (should be the roughly same)
+        toLedgerEffectiveTime = _ => trees.head.getEffectiveAt,
+      )
+
+      assertDisclosedContractsMetadata[Transaction](
+        streamType = "flatTransactionById",
+        expectedCount = 1,
+        streamResponses = Vector(flatById),
+        toCreateEvents = createdEvents,
+        toLedgerEffectiveTime = _.getEffectiveAt,
+      )
+
+      assertDisclosedContractsMetadata[TransactionTree](
+        streamType = "transactionTreeById",
+        expectedCount = 1,
+        streamResponses = Vector(treeById),
+        toCreateEvents = createdEvents,
+        toLedgerEffectiveTime = _.getEffectiveAt,
+      )
+    }
+  })
+
+  test(
+    "EDArchivedDisclosedContracts",
+    "The ledger rejects archived disclosed contracts",
+    allocate(Parties(2)),
+    enabled = _.explicitDisclosure,
+  )(implicit ec => { case Participants(Participant(ledger, owner, delegate)) =>
+    for {
+      testContext <- initializeTest(ledger, owner, delegate)
+
+      // Archive the disclosed contract
+      _ <- ledger.exercise(owner, testContext.delegatedCid.exerciseArchive())
+
+      // Exercise the choice using the now inactive disclosed contract
+      exerciseError <- testContext
+        .exerciseFetchDelegated(testContext.disclosedContract)
+        .mustFail("the contract is already archived")
+    } yield {
+      assertGrpcError(
+        exerciseError,
+        LedgerApiErrors.ConsistencyErrors.ContractNotFound,
+        None,
+        checkDefiniteAnswerMetadata = true,
+      )
+    }
+  })
+
+  test(
+    "EDDisclosedContractsArchiveRaceTest",
+    "Only when archival succeeds in a race between a normal exercise and one with disclosed contracts",
+    allocate(SingleParty, SingleParty),
+    enabled = _.explicitDisclosure,
+    repeated = 3,
+  )(implicit ec => {
+    case Participants(Participant(ledger1, party1), Participant(ledger2, party2)) =>
+      val attempts = 10
+
+      Future
+        .traverse((1 to attempts).toList) {
+          _ =>
+            for {
+              contractId <- ledger1.create(party1, Dummy(party1))
+
+              transactions <- ledger1.flatTransactionsByTemplateId(WithKey.id, party1)
+              create = createdEvents(transactions(1)).head
+              disclosedContract = createEventToDisclosedContract(create)
+
+              // Submit concurrently two consuming exercise choices (with and without disclosed contract)
+              party1_exerciseF = ledger1.exercise(party1, contractId.exerciseArchive())
+              party2_exerciseWithDisclosureF =
+                ledger2.submitAndWait(
+                  ledger2
+                    .submitAndWaitRequest(party2, contractId.exercisePublicChoice(party2).command)
```

Comment:
Any particular reason why you had to use submitAndWait here?

Reply:
I needed to create the submitAndWait request explicitly so I could inject the disclosed contract, if that's what you're asking. If you are looking at the difference to `.exercise`, they are both submitAndWait (but the exercise is waiting for the tx tree instead of the completion iirc)